aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -0,0 +1,507 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::borrow::Cow;
+use std::path::MAIN_SEPARATOR;
+use std::sync::atomic::Ordering;
+
+use super::{UploadObjectsContext, UploadObjectsInput};
+use async_channel::{Receiver, Sender};
+use aws_smithy_types::error::operation::BuildError;
+use blocking::Unblock;
+use futures_util::StreamExt;
+use walkdir::WalkDir;
+
+use crate::io::InputStream;
+use crate::operation::upload::UploadInputBuilder;
+use crate::operation::DEFAULT_DELIMITER;
+use crate::types::{FailedTransferPolicy, FailedUploadTransfer, UploadFilter};
+use crate::{error, types::UploadFilterItem};
+
+#[derive(Debug)]
+pub(super) struct UploadObjectJob {
+    key: String,
+    object: InputStream,
+}
+
+impl UploadObjectJob {
+    fn new(key: String, object: InputStream) -> Self {
+        Self { key, object }
+    }
+}
+
+pub(super) async fn list_directory_contents(
+    input: UploadObjectsInput,
+    work_tx: Sender<Result<UploadObjectJob, error::Error>>,
+) -> Result<(), error::Error> {
+    // Move a blocking I/O to a dedicated thread pool
+    let mut walker = Unblock::new(walker(&input).into_iter());
+
+    let default_filter = &UploadFilter::default();
+    let filter = input.filter().unwrap_or(default_filter);
+
+    while let Some(entry) = walker.next().await {
+        let job = match entry {
+            Ok(entry) => {
+                let metadata = tokio::fs::metadata(entry.path()).await?;
+                if !(filter.predicate)(&UploadFilterItem::new(entry.path(), metadata.clone())) {
+                    tracing::debug!("skipping object due to filter: {:?}", entry.path());
+                    continue;
+                }
+
+                let recursion_root_dir_path = input.source().expect("source set");
+                let entry_path = entry.path();
+                let relative_filename = entry_path
+                    .strip_prefix(recursion_root_dir_path)
+                    .expect("{entry_path:?} should be a path entry directly or indirectly under {recursion_root_dir_path:?}")
+                    .to_str()
+                    .expect("valid utf-8 path");
+                let object_key =
+                    derive_object_key(relative_filename, input.key_prefix(), input.delimiter())?;
+                let object = InputStream::read_from()
+                    .path(entry.path())
+                    .metadata(metadata)
+                    .build();
+
+                match object {
+                    Ok(object) => {
+                        tracing::info!(
+                            "uploading {relative_filename} with object key {object_key}..."
+                        );
+                        Ok(UploadObjectJob::new(object_key.into_owned(), object))
+                    }
+                    Err(e) => Err(crate::error::Error::from(BuildError::other(e))),
+                }
+            }
+            Err(e) => Err(crate::error::Error::from(BuildError::other(e))),
```

Comment:
Here is the walkdir [Error type](https://docs.rs/walkdir/latest/walkdir/struct.Error.html) which includes path and more.

I'm not sure `BuildError` is the right error here as it could be an IO error which isn't necessarily `InputInvalid` which is what this would get mapped to. I wonder if we need an `Other` error kind or if we just map this to `IOError` or `InputInvalid` depending on whether walkdir yields an IO error or not.